### PR TITLE
Add isEmpty method to the frontend compiler

### LINKED ISSUE #782: Add isEmpty method to frontend compiler
example:
```
box.R3[Int].isEmpty // check that value of R3 is undefined
```

### PR DESCRIPTION
**Pull request by darkdrag00n**
Fixes #782

Main contributions:
- Added the `isEmpty` method to the frontend compiler and transformed it to `LogicalNot(isDefined)` operation in the backend. This means that the cost of `isEmpty` operation is equal to `cost(LogicalNot) + cost(isDefined)`
- Added test cases for the newly added method

### CODE REVIEW

File: sigmastate/src/main/scala/sigmastate/types.scala [hunk 1]
```diff
@@ -1361,7 +1369,8 @@ object SOption extends STypeCompanion {
     FoldMethod,
     */
     MapMethod,
-    FilterMethod
+    FilterMethod,
+    IsEmptyMethod
```
Comment (aslesarenko) [inline sigmastate/src/main/scala/sigmastate/types.scala:1373]: we cannot add new methods to ErgoTree, because it will break the protocols.

**Review by aslesarenko — CHANGES_REQUESTED** [review]
The method should be added to SigmaCompiler without touching the interpreter and ErgoTree.
1) Please rebase on top of v5.x-restructure branch
2) in that brunch interpreter module is separate from sc module and thus the changes should be in sc. No changes in `interpreter` module.

**Comment by darkdrag00n** [comment]
@aslesarenko I am having trouble compiling the code in the `v5.x-restructure` branch. It fails to download the `scrypto` version `2.2.1-25-dafca54c-SNAPSHOT`. Any suggestions on what I can do?

Full error:
```
❯ sbt compile
[info] welcome to sbt 1.4.5 (Ubuntu Java 11.0.17)
[info] loading global plugins from /home/darkdrag00n/.sbt/1.0/plugins
[info] loading settings for project sigmastate-interpreter-build from plugins.sbt ...
[info] loading project definition from /home/darkdrag00n/freelancing/code/sigmastate-interpreter/project
[info] loading settings for project sigma from build.sbt ...
[info] set current project to sigma-state (in build file:/home/darkdrag00n/freelancing/code/sigmastate-interpreter/)
[info] Executing in batch mode. For better performance use sbt's shell
[info] Updating 
[info] Resolved  dependencies
[warn] 
[warn] 	Note: Unresolved dependencies path:
[info] compiling 30 Scala sources to /home/darkdrag00n/freelancing/code/sigmastate-interpreter/common/target/scala-2.12/classes ...
[info] Updating 
[info] Resolved  dependencies
[warn] 
[warn] 	Note: Unresolved dependencies path:
[warn] 8 feature warnings; re-run with -feature for details
[warn] one warning found
[error] sbt.librarymanagement.ResolveException: Error downloading org.scorexfoundation:scrypto_2.12:2.2.1-25-dafca54c-SNAPSHOT
[error]   Not found
[error]   Not found
[error]   not found: /home/darkdrag00n/.ivy2/localorg.scorexfoundation/scrypto_2.12/2.2.1-25-dafca54c-SNAPSHOT/ivys/ivy.xml
[error]   not found: https://repo1.maven.org/maven2/org/scorexfoundation/scrypto_2.12/2.2.1-25-dafca54c-SNAPSHOT/scrypto_2.12-2.2.1-25-dafca54c-SNAPSHOT.pom
[error]   not found: https://oss.sonatype.org/content/repositories/public/org/scorexfoundation/scrypto_2.12/2.2.1-25-dafca54c-SNAPSHOT/scrypto_2.12-2.2.1-25-dafca54c-SNAPSHOT.pom
[error] 	at lmcoursier.CoursierDependencyResolution.unresolvedWarningOrThrow(CoursierDependencyResolution.scala:258)
[error] 	at lmcoursier.CoursierDependencyResolution.$anonfun$update$38(CoursierDependencyResolution.scala:227)
[error] 	at scala.util.Either$LeftProjection.map(Either.scala:573)
[error] 	at lmcoursier.CoursierDependencyResolution.update(CoursierDependencyResolution.scala:227)
[error] 	at sbt.librarymanagement.DependencyResolution.update(DependencyResolution.scala:60)
[error] 	at sbt.internal.LibraryManagement$.resolve$1(LibraryManagement.scala:53)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$12(LibraryManagement.scala:103)
[error] 	at sbt.util.Tracked$.$anonfun$lastOutput$1(Tracked.scala:73)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$20(LibraryManagement.scala:116)
[error] 	at scala.util.control.Exception$Catch.apply(Exception.scala:228)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$11(LibraryManagement.scala:116)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$11$adapted(LibraryManagement.scala:97)
[error] 	at sbt.util.Tracked$.$anonfun$inputChangedW$1(Tracked.scala:219)
[error] 	at sbt.internal.LibraryManagement$.cachedUpdate(LibraryManagement.scala:130)
[error] 	at sbt.Classpaths$.$anonfun$updateTask0$5(Defaults.scala:3486)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] 	at sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] 	at sbt.Execute.work(Execute.scala:291)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error] 	at java.base/java.lang.Thread.run(Thread.java:829)
[error] sbt.librarymanagement.ResolveException: Error downloading org.scorexfoundation:scrypto_2.12:2.2.1-25-dafca54c-SNAPSHOT
[error]   Not found
[error]   Not found
[error]   not found: /home/darkdrag00n/.ivy2/localorg.scorexfoundation/scrypto_2.12/2.2.1-25-dafca54c-SNAPSHOT/ivys/ivy.xml
[error]   not found: https://repo1.maven.org/maven2/org/scorexfoundation/scrypto_2.12/2.2.1-25-dafca54c-SNAPSHOT/scrypto_2.12-2.2.1-25-dafca54c-SNAPSHOT.pom
[error]   not found: https://oss.sonatype.org/content/repositories/public/org/scorexfoundation/scrypto_2.12/2.2.1-25-dafca54c-SNAPSHOT/scrypto_2.12-2.2.1-25-dafca54c-SNAPSHOT.pom
[error] 	at lmcoursier.CoursierDependencyResolution.unresolvedWarningOrThrow(CoursierDependencyResolution.scala:258)
[error] 	at lmcoursier.CoursierDependencyResolution.$anonfun$update$38(CoursierDependencyResolution.scala:227)
[error] 	at scala.util.Either$LeftProjection.map(Either.scala:573)
[error] 	at lmcoursier.CoursierDependencyResolution.update(CoursierDependencyResolution.scala:227)
[error] 	at sbt.librarymanagement.DependencyResolution.update(DependencyResolution.scala:60)
[error] 	at sbt.internal.LibraryManagement$.resolve$1(LibraryManagement.scala:53)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$12(LibraryManagement.scala:103)
[error] 	at sbt.util.Tracked$.$anonfun$lastOutput$1(Tracked.scala:73)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$20(LibraryManagement.scala:116)
[error] 	at scala.util.control.Exception$Catch.apply(Exception.scala:228)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$11(LibraryManagement.scala:116)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$11$adapted(LibraryManagement.scala:97)
[error] 	at sbt.util.Tracked$.$anonfun$inputChangedW$1(Tracked.scala:219)
[error] 	at sbt.internal.LibraryManagement$.cachedUpdate(LibraryManagement.scala:130)
[error] 	at sbt.Classpaths$.$anonfun$updateTask0$5(Defaults.scala:3486)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] 	at sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] 	at sbt.Execute.work(Execute.scala:291)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error] 	at java.base/java.lang.Thread.run(Thread.java:829)
[error] (core-lib / update) sbt.librarymanagement.ResolveException: Error downloading org.scorexfoundation:scrypto_2.12:2.2.1-25-dafca54c-SNAPSHOT
[error]   Not found
[error]   Not found
[error]   not found: /home/darkdrag00n/.ivy2/localorg.scorexfoundation/scrypto_2.12/2.2.1-25-dafca54c-SNAPSHOT/ivys/ivy.xml
[error]   not found: https://repo1.maven.org/maven2/org/scorexfoundation/scrypto_2.12/2.2.1-25-dafca54c-SNAPSHOT/scrypto_2.12-2.2.1-25-dafca54c-SNAPSHOT.pom
[error]   not found: https://oss.sonatype.org/content/repositories/public/org/scorexfoundation/scrypto_2.12/2.2.1-25-dafca54c-SNAPSHOT/scrypto_2.12-2.2.1-25-dafca54c-SNAPSHOT.pom
[error] (update) sbt.librarymanagement.ResolveException: Error downloading org.scorexfoundation:scrypto_2.12:2.2.1-25-dafca54c-SNAPSHOT
[error]   Not found
[error]   Not found
[error]   not found: /home/darkdrag00n/.ivy2/localorg.scorexfoundation/scrypto_2.12/2.2.1-25-dafca54c-SNAPSHOT/ivys/ivy.xml
[error]   not found: https://repo1.maven.org/maven2/org/scorexfoundation/scrypto_2.12/2.2.1-25-dafca54c-SNAPSHOT/scrypto_2.12-2.2.1-25-dafca54c-SNAPSHOT.pom
[error]   not found: https://oss.sonatype.org/content/repositories/public/org/scorexfoundation/scrypto_2.12/2.2.1-25-dafca54c-SNAPSHOT/scrypto_2.12-2.2.1-25-dafca54c-SNAPSHOT.pom
```

The build on `develop` branch works because it looks for version `2.1.10`

**Comment by darkdrag00n** [comment]
Updating the version of scrypto from `2.2.1-25-dafca54c-SNAPSHOT` to `2.2.1` worked.

**Comment by aslesarenko** [comment]
@darkdrag00n, Sigma codebase is under major refactoring now, and project structure has been changed significantly.

 The new features should be added on top of the latest branch https://github.com/ScorexFoundation/sigmastate-interpreter/pull/833. It has some dependencies which are not published and still under review by @kushti.

I'm currently publish such dependencies locally.

You need to clone this branch https://github.com/input-output-hk/scrypto/pull/98
and do `sbt publishLocal`. 

See the console output to find the correct version (something like 2.2.1-25-dafca54c-SNAPSHOT, but may differ).
Then use that version in build.sbt of Sigma.

File: sigmastate/src/main/scala/sigmastate/types.scala (hunk 1 above)
Comment (darkdrag00n) [reply sigmastate/src/main/scala/sigmastate/types.scala]: Removed.

**Comment by darkdrag00n** [comment]
@aslesarenko I've limited the changes to the `sc` module now. Also, changed the target branch. Ready for review now.

**Review by aslesarenko — APPROVED** [review]
LGTM

**Pull request closed without merging**